### PR TITLE
`NoiseModel` example for `mitigate_with_zne`

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Adding `NoiseModel` to a device should work properly with `mitigate_with_zne`

**Description of the Change:** Added an example for the docs and improved the error message.

**Benefits:** Users know how to use noise models with `mitigate_with_zne`

**Possible Drawbacks:**

**Related GitHub Issues:** [sc-72624]

### CODE REVIEW

File: tests/transforms/test_mitigate.py [hunk 1]
```diff
@@ -208,6 +209,34 @@ def original_qnode(inputs):
...
+            zne_qnode = qml.transforms.mitigate_with_zne(
+                noise_qnode, [1, 2, 3], fold_global, richardson_extrapolate
+            )
+            zne_qnode()
+
+        noise_qnode = qml.QNode(circuit, device=qml.add_noise(dev_ideal, noise_model))
+        zne_qnode = qml.transforms.mitigate_with_zne(
+            noise_qnode, [1, 2, 3], fold_global, richardson_extrapolate
+        )
+        assert qml.math.allclose(zne_qnode(), 0.3984378845598476, atol=1e-2)
```
Comment (KetpuntoG) [inline tests/transforms/test_mitigate.py:238]: How has been calculated this value?

**Review by KetpuntoG — COMMENTED** [review]
Looking good! 💪 Just two comments

File: pennylane/transforms/mitigate.py [hunk 2]
```diff
@@ -409,16 +415,19 @@ def mitigate_with_zne(
...
 
     .. code-block:: python3
 
         import pennylane as qml
 
-        noise_strength = 0.05
-
         dev = qml.device("default.mixed", wires=2)
-        dev = qml.transforms.insert(dev, qml.AmplitudeDamping, noise_strength)
+
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/transforms/mitigate.py:425]: You will have to wait for [this](https://github.com/PennyLaneAI/pennylane/pull/6321) to go in for this to work!

File: tests/transforms/test_mitigate.py (hunk 1 above)
Comment (obliviateandsurrender) [reply tests/transforms/test_mitigate.py:238]: I computed it manually via `mitiq`.

File: pennylane/transforms/mitigate.py [hunk 3]
```diff
@@ -446,7 +456,8 @@ def mitigate_with_zne(
             folding=fold_global,
             extrapolate=poly_extrapolate,
             extrapolate_kwargs={'order': 2})
```
Comment (isaacdevlugt) [inline pennylane/transforms/mitigate.py:458]:
```suggestion
            extrapolate_kwargs={'order': 2}
        )
```

File: pennylane/transforms/mitigate.py [hunk 4]
```diff
@@ -446,7 +456,8 @@ def mitigate_with_zne(
...
-        @qnode(dev)
+        @partial(qml.compile, basis_set = ["RY", "CZ"])
```
Comment (isaacdevlugt) [inline pennylane/transforms/mitigate.py:459]:
```suggestion
        @partial(qml.transforms.decompose, basis_set = ["RY", "CZ"])
```

File: tests/transforms/test_mitigate.py (hunk 1 above)
Comment (isaacdevlugt) [reply tests/transforms/test_mitigate.py:238]: @obliviateandsurrender do you have the source code for that? Just curious :)

**Review by isaacdevlugt — APPROVED** [review]
Thanks @obliviateandsurrender! Looks really nice. I'm wondering if we can add a `note` directive somewhere that explains the error message in a bit more detail. E.g., 

```
.. note ::

    If a `NoiseModel` contains a :class:`~.operation.Channel` instance and `add_noise` is applied to the circuit (not the device), `mitigate_with_zne` will try to take the adjoint of the channel operation when folding, which is not currently supported. To avoid this, please use `add_noise` on the device instead of the circuit.
```

What do you think?

File: pennylane/transforms/mitigate.py [hunk 5]
```diff
@@ -210,6 +210,12 @@ def qfunc(op):
     # Treat all circuits as lists of operations, build new tape in the end
 
     base_ops = circuit.expand().copy(copy_operations=True).operations
+    if any((isinstance(op, qml.operation.Channel) for op in base_ops)):
+        raise ValueError(
+            "Circuits with channels cannot be folded with mitigate_with_zne. "
+            "To use zero-noise extrapolation on the circuit with channel noise, "
+            "consider adding the noise on the device rather than the circuit."
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/transforms/mitigate.py:217]: I don't want to put "add_noise" here since this error could be raised in a non-noise model scenario as well.

**Comment by isaacdevlugt** [comment]
Looks good to me!

File: tests/transforms/test_mitigate.py (hunk 1 above)
Comment (soranjh) [reply tests/transforms/test_mitigate.py:238]: Please add some information that allows reproducing the reference value, ideally the full code or the important part of it if too long.

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @obliviateandsurrender, left a few comments.

File: tests/transforms/test_mitigate.py (hunk 1 above)
Comment (obliviateandsurrender) [reply tests/transforms/test_mitigate.py]: I personally obtained it with some local changes to `mitiq` to extract the results of their `global_folding` functionality and then feed it to their `RichardsonFactory` method, while defining the circuit in `cirq` as below:
```py
circuit = cirq.Circuit()
q0, q1 = cirq.LineQubit.range(2)
circuit.append(cirq.rx(1.23).on(q0))
circuit.append(cirq.rz(0.45).on(q1))

z = np.diag([1, -1])
hamiltonian = np.kron(z, z)
noiseless_executor = lambda circ: np.real(
    np.trace(cirq.DensityMatrixSimulator().simulate(circ).final_density_matrix @ hamiltonian)
)
executor_with_noise = lambda circ: noiseless_executor(
    circ.with_noise(cirq.amplitude_damp(0.01))
)
```
An easier alternative could be to simply use mitiq's own function with our transform:
```py
>>> qml.transforms.mitigate_with_zne(
    noise_qnode, [1, 2, 3], 
    mitiq.zne.scaling.fold_global, mitiq.zne.inference.RichardsonFactory.extrapolate
)()
0.3984378845760577
```
I have added that in the comment.

File: pennylane/transforms/mitigate.py [hunk 6]
```diff
@@ -450,9 +466,9 @@ def mitigate_with_zne(
             scale_factors=[1., 2., 3.],
             folding=fold_global,
             extrapolate=poly_extrapolate,
-            extrapolate_kwargs={'order': 2})
+        )
```
Comment (KetpuntoG) [inline pennylane/transforms/mitigate.py:469]: I get this error when running the example:
`TypeError: mitigate_with_zne() missing 1 required positional argument: 'extrapolate'`

File: pennylane/transforms/mitigate.py [hunk 7]
```diff
@@ -450,17 +466,18 @@ def mitigate_with_zne(
             scale_factors=[1., 2., 3.],
             folding=fold_global,
             extrapolate=poly_extrapolate,
-            extrapolate_kwargs={'order': 2})
+            extrapolate_kwargs={"order":2},
```
Comment (obliviateandsurrender) [inline pennylane/transforms/mitigate.py:469]:
```suggestion
            extrapolate_kwargs={"order" : 2},
```

File: pennylane/transforms/mitigate.py [hunk 8]
```diff
@@ -450,17 +466,18 @@ def mitigate_with_zne(
             scale_factors=[1., 2., 3.],
             folding=fold_global,
             extrapolate=poly_extrapolate,
-            extrapolate_kwargs={'order': 2})
+            extrapolate_kwargs={"order" : 2},
```
Comment (obliviateandsurrender) [inline pennylane/transforms/mitigate.py:469]:
```suggestion
            extrapolate_kwargs={'order' : 2},
```

File: pennylane/transforms/mitigate.py (hunk 6 above)
Comment (obliviateandsurrender) [reply pennylane/transforms/mitigate.py:470]: Should be good now.

File: pennylane/transforms/mitigate.py [hunk 9]
```diff
@@ -433,7 +449,7 @@ def mitigate_with_zne(
 
     .. code-block:: python3
 
-        import numpy as np
+        import pennylane.numpy as np
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/transforms/mitigate.py]: Reverted it.

**Review by KetpuntoG — APPROVED** [review]
🚀

File: pennylane/transforms/mitigate.py [hunk 10]
```diff
@@ -215,6 +215,12 @@ def qfunc(op):
     # Generate base_circuit without measurements
     # Treat all circuits as lists of operations, build new tape in the end
     base_ops = circuit.operations
+    if any((isinstance(op, qml.operation.Channel) for op in base_ops)):
+        raise ValueError(
+            "Circuits containing quantum channels cannot be folded with mitigate_with_zne. "
+            "To use zero-noise extrapolation on the circuit with channel noise, "
+            "consider adding the noise on the device rather than the circuit."
```
Comment (soranjh) [inline pennylane/transforms/mitigate.py:222]:
```suggestion
            "please include the noise on the device rather than the circuit."
```

Comment (soranjh) [reply pennylane/transforms/mitigate.py:222]: "consider" seems optional, but the user "has to" add the noise to the device, correct?

File: tests/transforms/test_mitigate.py [hunk 11]
```diff
@@ -208,6 +209,49 @@ def original_qnode(inputs):
...
+            qml.RZ(0.45, wires=1)
+            return qml.expval(qml.Z(0) @ qml.Z(1))
+
+        noise_qnode = qml.QNode(circuit, device=qml.add_noise(dev_ideal, noise_model))
+        zne_qnode = qml.transforms.mitigate_with_zne(
+            noise_qnode, [1, 2, 3], fold_global, richardson_extrapolate
+        )
+
+        # following result has been obtained manually and also by using
+        # mitiq.zne.scaling.fold_global and mitiq.zne.inference.RichardsonFactory.extrapolate
```
Comment (soranjh) [inline tests/transforms/test_mitigate.py:230]: Will be great to add your code:

```
# qml.transforms.mitigate_with_zne(
#     noise_qnode, [1, 2, 3], 
#     mitiq.zne.scaling.fold_global,
#     mitiq.zne.inference.RichardsonFactory.extrapolate)()
```

**Review by soranjh — APPROVED** [review]
Thanks @obliviateandsurrender, left some final minor comments.

**Pull request merged**